lib/rouge/langspec.rb

```diff
@@ -0,0 +1,126 @@
+module Rouge
+  class LangSpec
+    def to_s; inspect; end
+
+    def self.delegate(m, &default)
+      define_method(m) { find_prop(m) || (default && default.call) }
+    end
+
+    def self.load_and_delegate(m)
+      define_method(m) { |*a, &b| lexer_class.send(m, *a, &b) }
+    end
+
+    delegate(:tag)
+    delegate(:title)
+    delegate(:desc)
+    delegate(:option_docs) { {} }
+    delegate(:aliases) { [] }
+    delegate(:filenames) { [] }
+    delegate(:mimetypes) { [] }
+    delegate(:detectable?)
+
+    load_and_delegate :lex
+    load_and_delegate :continue_lex
+    load_and_delegate :new
+
+    def load!
+      # noop
+    end
+
+    # overridden with a `def self.detect?` in
+    # the cache file, but some classes don't
+    # define it. this is the default implementation.
+    def detect?(analyzer)
+      false
+    end
+
+    # for compat reasons, a LangSpec is considered == to the
+    # lexer class itself.
+    def ==(other)
+      return true if other.respond_to?(:tag) && other.tag == self.tag
+      super
+    end
+
+    ### abstract methods ###
+
+    def find_prop(prop)
+      raise 'abstract'
+    end
+
+    def lexer_class
+      raise 'abstract'
+    end
+
+    def name
+      raise 'abstract'
+    end
+  end
+
+  class BakedLangSpec < LangSpec
+    def inspect
+      "#<BakedLangSpec #{@lexer_class}>"
+    end
+
+    attr_reader :lexer_class
+    def initialize(lexer_class)
+      @lexer_class = lexer_class
+    end
+
+    delegate :name
+    delegate :demo_file
+    delegate :demo
+
+    def detect?(text)
+      @lexer_class.detect?(text)
+    end
+
+    def find_prop(prop)
+      @lexer_class.send(prop)
+    end
+  end
+
+  class CachedLangSpec < LangSpec
+    def inspect
+      "#<CachedLangSpec:#{@const_name} #{@source_file}>"
+    end
+
+    def initialize(source_file, const_name, &b)
+      @const_name = const_name
+      @source_file = source_file
+      instance_eval(&b)
+    end
+
+    def find_prop(prop)
+      instance_variable_get(:"@#{prop}")
```

Comment:
In a strict sense regarding memory allocation and GC trigger, using (`private` ?) `attr_reader` would be better here to avoid unnecessary interpolation and conversion to Symbol. If `prop` can be both String or Symbol, you could alternatively dispatch a `:send` call to the private attr_reader:
```ruby
def find_prop(prop)
  send prop
end
```

Reply:
This is not equivalent to `attr_reader`, unfortunately - it's an instance method.

Reply:
Also, I should mention - my original strategy used a Hash, so this would have been `@props[prop]`. This approach saved on memory in all cases.

Reply:
> This is not equivalent to `attr_reader`, unfortunately - it's an instance method.

Perhaps my example wasn't clear enough.. 
What I meant is...
*Given the following two classes:*
```ruby
class Foo
  attr_reader :arg

  def initialize(arg, &block)
    @arg = arg
    instance_eval(&block)
  end

  def find_prop(prop)
    send prop
  end

  private

  attr_reader :tag, :title, :desc, :filenames, :mimetypes
end

class Bar < Foo
  def find_prop(prop)
    instance_variable_get("@#{prop}")
  end
end
```
With the following instances:
```ruby
  foo = Foo.new('alpha') do
    @tag = "abap"
    @title = "ABAP"
    @desc = "SAP - Advanced Business Application Programming"
    @filenames = ["*.abap"]
    @mimetypes = ["text/x-abap"]
    @detectable = false
  end

  bar = Bar.new('neta') do
    @tag = "abap"
    @title = "ABAP"
    @desc = "SAP - Advanced Business Application Programming"
    @filenames = ["*.abap"]
    @mimetypes = ["text/x-abap"]
    @detectable = false
  end
```
And a test script:
```ruby
def process(obj)
  [:arg, :tag, :title, :desc, :filenames, :mimetypes].map { |prop| obj.find_prop(prop) }
end
```
Then my hypothesis is that:
- `process(foo).hash == process(bar).hash`,
- but memory allocation from `process(foo)` **is lesser than** memory allocation from `process(bar)`